File: go-backend-interview-prep/07_io_files_networking/01_io_reader_writer_basics/main.py
"""IO Reader/Writer basics -- Python equivalent of the Go example.

Python uses file-like objects with .read() / .write() methods.
io.StringIO and io.BytesIO are the in-memory equivalents.
"""

import io
import shutil


class CountingReader:
    """Wraps a file-like object and counts bytes read."""

    def __init__(self, reader):
        self._reader = reader
        self.bytes_read = 0

    def read(self, n=-1):
        data = self._reader.read(n)
        self.bytes_read += len(data)
        return data


def main() -> None:
    # --- Example 1: io.StringIO (like strings.NewReader) ---
    print("=== io.StringIO ===")
    r = io.StringIO("hello, Go IO!")

    while True:
        chunk = r.read(5)
        if not chunk:
            break
        print(f"  read {len(chunk)} chars: {chunk!r}")

    # --- Example 2: io.StringIO as read+write buffer ---
    print("\n=== io.StringIO (buffer) ===")
    buf = io.StringIO()
    buf.write("hello ")
    buf.write("buffer")
    print("  buffer contents:", buf.getvalue())

    buf.seek(0)
    out = buf.read(5)
    print(f"  read {len(out)} chars: {out!r}")

    # --- Example 3: shutil.copyfileobj (like io.Copy) ---
    print("\n=== shutil.copyfileobj ===")
    src = io.BytesIO(b"streaming data via copyfileobj")
    dst = io.BytesIO()

    shutil.copyfileobj(src, dst)
    content = dst.getvalue()
    print(f"  copied {len(content)} bytes: {content.decode()!r}")

    # --- Example 4: custom CountingReader ---
    print("\n=== Custom CountingReader ===")
    cr = CountingReader(io.StringIO("count these bytes"))
    result = cr.read()
    print(f"  content: {result!r}")
    print(f"  total bytes read: {cr.bytes_read}")


if __name__ == "__main__":
    main()

File: go-backend-interview-prep/07_io_files_networking/01_io_reader_writer_basics/main.go
package main

import (
	"bytes"
	"fmt"
	"io"
	"strings"
)

// CountingReader wraps a Reader and counts bytes read.
type CountingReader struct {
	reader    io.Reader
	BytesRead int
}

func (cr *CountingReader) Read(p []byte) (int, error) {
	n, err := cr.reader.Read(p)
	cr.BytesRead += n
	return n, err
}

func main() {
	// --- Example 1: strings.NewReader ---
	fmt.Println("=== strings.NewReader ===")
	r := strings.NewReader("hello, Go IO!")
	buf := make([]byte, 5)

	for {
		n, err := r.Read(buf)
		if n > 0 {
			fmt.Printf("  read %d bytes: %q\n", n, buf[:n])
		}
		if err == io.EOF {
			break
		}
		if err != nil {
			fmt.Println("  error:", err)
			break
		}
	}

	// --- Example 2: bytes.Buffer (Reader + Writer) ---
	fmt.Println("\n=== bytes.Buffer ===")
	var b bytes.Buffer
	b.WriteString("hello ")
	b.WriteString("buffer")
	fmt.Println("  buffer contents:", b.String())

	// Read from buffer
	out := make([]byte, 5)
	n, _ := b.Read(out)
	fmt.Printf("  read %d bytes: %q\n", n, out[:n])

	// --- Example 3: io.Copy ---
	fmt.Println("\n=== io.Copy ===")
	src := strings.NewReader("streaming data via io.Copy")
	var dst bytes.Buffer

	written, err := io.Copy(&dst, src)
	if err != nil {
		fmt.Println("  error:", err)
	}
	fmt.Printf("  copied %d bytes: %q\n", written, dst.String())

	// --- Example 4: custom Reader ---
	fmt.Println("\n=== Custom CountingReader ===")
	cr := &CountingReader{reader: strings.NewReader("count these bytes")}
	var result bytes.Buffer
	io.Copy(&result, cr)
	fmt.Printf("  content: %q\n", result.String())
	fmt.Printf("  total bytes read: %d\n", cr.BytesRead)
}
